Rubberduck.Refactorings/Abstract/IBaseRefactoring.cs

```diff
@@ -0,0 +1,39 @@
+﻿using Rubberduck.Parsing.Rewriter;
+
+namespace Rubberduck.Refactorings
+{
+    /// <summary>
+    /// The heart of a refactoring: this part of the refactoring performs the actual transformation of the code once all necessary information has been gathered.
+    /// </summary>
+    /// <typeparam name="TModel">The model used by the refactoring containing all information needed to specify what to do.</typeparam>
+    public interface IBaseRefactoring<in TModel> where TModel : class, IRefactoringModel
+    {
+        /// <summary>
+        /// Performs the actual refactoring based on the parameters specified in the model.
+        /// </summary>
+        /// <param name="model">The model specifying all parameters of the refactoring</param>
+        void Refactor(TModel model);
+    }
+
+    public interface IRefactoringPreviewProvider<in TModel>
+        where TModel : class, IRefactoringModel
+    {
+        /// <summary>
+        /// Returns some preview of the refactored code.
+        /// </summary>
+        /// <param name="model">The model used by the refactoring containing all information needed to specify what to do.</param>
+        /// <returns>Preview of the refactored code</returns>
+        string Preview(TModel model);
+    }
+
+    public interface ICodeOnlyBaseRefactoring<in TModel> : IBaseRefactoring<TModel>
+        where TModel : class, IRefactoringModel
+    {
+        /// <summary>
+        /// Performs the refactoring according to the model and using the provided rewrite session.
+        /// </summary>
+        /// <param name="model">The model specifying all parameters of the refactoring</param>
+        /// <param name="rewriteSession">Rewrite session used to manipulate the code (Does not get executed.)</param>
+        void Refactor(TModel model, IRewriteSession rewriteSession);
```

Comment:
I'm understanding that `ICodeOnlyBaseRefactoring` is only intended for refactorings that can be safely used within other refactorings.  We do not currently have refactorings that would use it, but the `MoveMember` refactoring may be useful within other (future) refactorings.  If so, passing only a rewriteSession and not an `IRewritingManager` in the `Refactor` interface member may disqualify it's use in this way.  

The `MoveMember` refactoring leverages our rewriting tools by using a _scratchpad_ rewriteSession different from the _primary_ rewriteSession.  The _primary_ rewriteSession would be the one passed by `Refactor(...)` and it modifies the `MoveMemberRefactoring's` Source and Destination module content.  The _scratchpad_ rewriteSession is used to generate the code that is inserted in the Destination module.  The _scratchpad_ rewriteSession operates on the declarations-to-move in the Source module.  These modifications include renaming if there will be name conflicts when moved to the Destination and removing `MemberAccess` contexts in cases where the moved members contain module qualified references to the Destination's members and fields.  This _scratchpad_ session then generates the new/modified content via `GetText()` and is abandoned (`TryRewrite()` is never called).  In the _primary_ rewrite session, the moved members are removed from the Source module and the moved content text block is inserted in the Destination when the _primary_ session's `TryRewrite()` is called.

Reply:
I do not see the problem here. In the end, all the described refactoring does is make code changes on one rewrite session. Nothing prevents it from having the `IRewritingManager` constructor injected from which it can get the _scratchpad_.

Reply:
To elaborate on this a bit further. What you describe with the _scratchpad_ would actually be a typical implementation of the new `IRefactoringPreviewProvider`. The entire refactoring could actually be split into an `ICodeOnlyRefactoringAction` that does the refactoring, a wrapper for the preview using it to produce a preview and the actual refactoring generating a new model based on the information gathered from the preview, which then gets applied by the refactoring action.